datafusion/core/src/physical_optimizer/pruning.rs

```diff
@@ -553,22 +553,33 @@ fn is_compare_op(op: Operator) -> bool {
 
 /// replaces a column with an old name with a new name in an expression
 fn rewrite_column_expr(
-    expr: &Expr,
+    e: Expr,
     column_old: &Column,
     column_new: &Column,
 ) -> Result<Expr> {
-    let expressions = utils::expr_sub_expressions(expr)?;
-    let expressions = expressions
-        .iter()
-        .map(|e| rewrite_column_expr(e, column_old, column_new))
-        .collect::<Result<Vec<_>>>()?;
-
-    if let Expr::Column(c) = expr {
-        if c == column_old {
-            return Ok(Expr::Column(column_new.clone()));
+    struct ColumnReplacer<'a> {
+        old: &'a Column,
+        new: &'a Column,
+    }
+
+    impl<'a> ExprRewriter for ColumnReplacer<'a> {
+        fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+            if let Expr::Column(c) = &expr {
+                if c == self.old {
+                    Ok(Expr::Column(self.new.clone()))
+                } else {
+                    Ok(expr)
+                }
+            } else {
+                Ok(expr)
+            }
```

Comment:
minor nit: this can be simplified:

```suggestion
            match expr {
                Expr::Column(c) if c == *self.old => Ok(Expr::Column(self.new.clone())),
                _ => Ok(expr)
            }
```